### ISSUE TITLE
IIS concurrency issue

### ISSUE THREAD

**Issue by caracostea** [SpiderLabs/ModSecurity#1128]
On my test server I've noticed the following behavior:
The more requests I throw at it, the higher the response times become, but without topping the CPU, memory, bandwidth, etc.
The CPU usage seems to become limited to ~25% and from there on, the response times keep getting higher.
I went through a bit of code from the IIS module and found that the code inside `CMyHttpModule::OnSendResponse`, `CMyHttpModule::OnPostEndRequest` and `CMyHttpModule::OnBeginRequest` is wrapped inside 
```
EnterCriticalSection(&m_csLock); 
...
LeaveCriticalSection(&m_csLock);
```

That, coupled with the singleton approach of `CMyHttpModuleFactory::GetHttpModule()` and the fact that `m_csLock` is a variable of `CMyHttpModule` only initialized in its constructor (called only once), means that all the requests which arrive on different threads are put on hold and executed one by one.

Am I missing something?

**Comment by zimmerle** [comment]
Hi @caracostea, 

It seems that your comment is pertinent. If you are thinking in patch this code, to provide something more clever i would suggest to have a look at the libmodsecurity. We may need a new connector for IIS using the libmodsecurity. See here:

https://github.com/SpiderLabs/ModSecurity/tree/libmodsecurity

**Comment by caracostea** [comment]
@zimmerle I need to patch this one real quick and I see that work on IIS connector has not started so far. What is the reason for which those code blocks must be placed in critical sections? Which are the resources modified cross-thread? I looked at the code (not in depth, I admit) and failed to spot them.

**Comment by zimmerle** [comment]
Hi @caracostea,

It seems it is that way since the very first commit I am contacting the original developer of this module to see if he can help us.

**Comment by caracostea** [comment]
@zimmerle Thanks, that'd be great! Also, consider reopening the issue. Maybe someone else would also like to contribute.
